File: src/main/java/com/projectswg/forwarder/services/client/ClientProtocolService.java
package com.projectswg.forwarder.services.client;

import com.projectswg.forwarder.intents.client.SonyPacketInboundIntent;
import com.projectswg.forwarder.resources.networking.data.ProtocolStack;
import com.projectswg.forwarder.resources.networking.packets.ClientNetworkStatusUpdate;
import com.projectswg.forwarder.resources.networking.packets.KeepAlive;
import com.projectswg.forwarder.resources.networking.packets.Packet;
import com.projectswg.forwarder.resources.networking.packets.ServerNetworkStatusUpdate;
import me.joshlarson.jlcommon.control.IntentHandler;
import me.joshlarson.jlcommon.control.IntentMultiplexer;
import me.joshlarson.jlcommon.control.IntentMultiplexer.Multiplexer;
import me.joshlarson.jlcommon.control.Service;

public class ClientProtocolService extends Service {
	
	private static final int GALACTIC_BASE_TIME	= 1323043200;
	
	private final IntentMultiplexer multiplexer;
	
	public ClientProtocolService() {
		this.multiplexer = new IntentMultiplexer(this, ProtocolStack.class, Packet.class);
	}
	
	@IntentHandler
	private void handleSonyPacketInboundIntent(SonyPacketInboundIntent spii) {
		multiplexer.call(spii.getStack(), spii.getPacket());
	}
	
	@Multiplexer
	private void handleClientNetworkStatus(ProtocolStack stack, ClientNetworkStatusUpdate update) {
		ServerNetworkStatusUpdate serverNet = new ServerNetworkStatusUpdate();
		serverNet.setClientTickCount((short) update.getTick());
		serverNet.setServerSyncStampLong((int) (System.currentTimeMillis()-GALACTIC_BASE_TIME));
		serverNet.setClientPacketsSent(update.getSent());
		serverNet.setClientPacketsRecv(update.getRecv());
		serverNet.setServerPacketsSent(stack.getTxSequence());
		serverNet.setServerPacketsRecv(stack.getRxSequence());
		stack.send(serverNet);
	}
	
	@Multiplexer
	private void handleKeepAlive(ProtocolStack stack, KeepAlive keepAlive) {
		stack.send(new KeepAlive());
	}
	
}

File: src/main/kotlin/com/projectswg/forwarder/services/client/ClientProtocolService.kt
package com.projectswg.forwarder.services.client

import com.projectswg.forwarder.intents.SonyPacketInboundIntent
import com.projectswg.forwarder.resources.networking.data.ProtocolStack
import com.projectswg.forwarder.resources.networking.packets.ClientNetworkStatusUpdate
import com.projectswg.forwarder.resources.networking.packets.KeepAlive
import com.projectswg.forwarder.resources.networking.packets.Packet
import com.projectswg.forwarder.resources.networking.packets.ServerNetworkStatusUpdate
import me.joshlarson.jlcommon.control.IntentHandler
import me.joshlarson.jlcommon.control.IntentMultiplexer
import me.joshlarson.jlcommon.control.IntentMultiplexer.Multiplexer
import me.joshlarson.jlcommon.control.Service

class ClientProtocolService : Service() {
	
	private val multiplexer: IntentMultiplexer = IntentMultiplexer(this, ProtocolStack::class.java, Packet::class.java)
	
	@IntentHandler
	private fun handleSonyPacketInboundIntent(spii: SonyPacketInboundIntent) {
		multiplexer.call(spii.stack, spii.packet)
	}
	
	@Multiplexer
	private fun handleClientNetworkStatus(stack: ProtocolStack, update: ClientNetworkStatusUpdate) {
		val serverNet = ServerNetworkStatusUpdate()
		serverNet.clientTickCount = update.tick.toShort()
		serverNet.serverSyncStampLong = (System.currentTimeMillis() - GALACTIC_BASE_TIME).toInt()
		serverNet.clientPacketsSent = update.sent
		serverNet.clientPacketsRecv = update.recv
		serverNet.serverPacketsSent = stack.txSourceSequence
		serverNet.serverPacketsRecv = stack.rxSourceSequence
		stack.send(serverNet)
	}
	
	@Multiplexer
	private fun handleKeepAlive(stack: ProtocolStack, keepAlive: KeepAlive) {
		stack.send(KeepAlive())
	}
	
	companion object {
		
		private const val GALACTIC_BASE_TIME = 1323043200
		
	}
	
}
